0015-automatic-cluster-upgrades/0015-automatic-cluster-upgrades.md

```diff
@@ -0,0 +1,139 @@
+
+## Concepts
+
+
+### Cluster upgrade entity
+
+There are different reason why we are interested in having an object to define the upgrade intent. 
+
+- As we define in the user stories, we/customers would like to schedule upgrades on certain date from running version to a specific one. That way customer can plan upgrades ahead and so do we.
+
+- We have a registry of upgrades done to a cluster for auditing and tracking purposes.
+
+- We can save cluster upgrade status displaying if something goes wrong or not.
+
+- In theory we can include apps upgrades together or separately.
+
+### Upgrade policy entity
+
+Our customer can have a different set of constraints when it comes to allow changes in their clusters. The upgrade policy would define those and then the clusters can rely on a policy to enforce the desired behavior.
+
+The upgrade policy would be possible to be referenced from different cluster resources (Cluster,Machine Pool, KubeadmControlPlane, Apps,...) letting controllers to decide if the resource would be or not reconciled.
+
+The upgrade policy would contain
+
+- A way to define dates or times where clusters (or relatives) can be modified.
+
+- A way to define dates or times where clusters (or relatives) cannot be modified.
+
+## Technical aspects
+
+We envision aforementioned concepts would become custom resource definitions and there will be additional controller that would enhance and ensure the correct behavior.
+
+### Resources
+
+#### ClusterUpgrade
+
+Example resource:
+
+```yaml
+apiVersion: 
+kind: ClusterUpgrade
+metadata:
+  name: upgrade-to-release-v20
+  namespace: default
+spec:
+  clusterRef:
+    apiVersion: cluster.x-k8s.io/v1alpha3
+    kind: Cluster
+    name: foo01
+  kubernetesVersion: 1.19.8
+```
+
+#### UpgradePolicy
+
+Example resource:
+
+```yaml
+apiVersion: 
+kind: UpgradePolicy
+metadata:
+  name: working-hours-except-month-start
+  namespace: default
+spec:
+  maintenanceWindows:
+  - name: working-hours
+    startTime: 2021-03-10T07:00:00Z
+    endTime: 2021-03-10T17:00:00Z
+    recurrence: FREQ=WEEKLY;BYDAY=MO,TU,WE,TH,FR
+  maintenanceExclusions:
+  - name: month-start
+    startTime: 2021-03-01T00:00:00Z
+    endTime: 2021-03-10T23:59:59Z
+    recurrence: FREQ=MONTHLY
+```
+
+This would allow upgrades to happen between 07:00 UTC and 17:00 UTC on Monday-Friday, except for days between 1st and 10th of a month.
+
+### Controllers
+
+#### Cluster upgrade executor
+
+Thee will be a cluster that based on the `Cluster Upgrade` CRs will initiate the upgrades changing the labels on the specific CR(s), it could include Apps. 
+
+This operator can also check on the `Upgrade Policy` if the cluster can be done or not, and in case it cannot happen alert or set an appropriate status on the object status.
```

Comment:
```suggestion
There will be an operator that based on the `Cluster Upgrade` CRs will initiate the upgrades changing the labels on the specific CR(s), it could include Apps at some point though ideally would detach apps. totally from the cluster and. apps metadata and App Platform enforce the requirements. 

This operator can also check on the `Upgrade Policy` to verify the upgrade fulfil or not the criteria, and in case it does not, alert or set an appropriate status on the object status.
```